### PR TITLE
split index and block data

### PR DESCRIPTION
**Pull request by bobotu**
Use `.idx` file to store data of index. Also add some required meta into index file, so that `OpenTable` doesn't need to load block data.

PTAL @coocood

### CODE REVIEW

File: table/builder.go [hunk 1]
```diff
@@ -273,10 +298,40 @@ func (b *Builder) EstimateSize() int {
...
+	}
+	binary.BigEndian.PutUint64(encodeBuf[:], ts)
+	if err := b.w.Append(encodeBuf[:]); err != nil {
+		return err
+	}
+
+	binary.BigEndian.PutUint64(encodeBuf[:], uint64(b.writtenLen))
+	b.buf = append(b.buf, encodeBuf[:]...)
+
+	b.buf = append(b.buf, u32ToBytes(uint32(len(b.smallest)))...)
```
Comment (coocood) [inline table/builder.go]: I think it's time we add a field flag for each index field, so later adding a new field would be compatible.

**Comment by bobotu** [comment]
Added the `metaEncoder` and `metaDecoder` PTAL @coocood

File: table/builder.go [hunk 2]
```diff
@@ -424,3 +437,69 @@ func (b *Builder) compressData(data []byte) ([]byte, error) {
 	}
 	return nil, errors.New("Unsupported compression type")
 }
+
+// type use 2 bits, now we have two unused type.
+const (
+	typeUint    = 0
```
Comment (coocood) [inline table/builder.go]: Can we remove the type?
For each field, the type is known.

Comment (bobotu) [reply table/builder.go]: We have to support `metaDecoder.skip`

File: table/builder.go [hunk 3]
```diff
@@ -307,67 +323,54 @@ func (b *Builder) Finish() error {
...
-	b.buf = append(b.buf, b.smallest...)
-	b.buf = append(b.buf, u32ToBytes(uint32(len(b.biggest)))...)
-	b.buf = append(b.buf, b.biggest...)
+	encoder := metaEncoder{b.buf}
 
-	b.buf = append(b.buf, u32ToBytes(uint32(len(b.baseKeysEndOffs)))...)
-	b.buf = append(b.buf, u32SliceToBytes(b.baseKeysEndOffs)...)
-	b.buf = append(b.buf, b.baseKeysBuf...)
-	b.buf = append(b.buf, u32SliceToBytes(b.blockEndOffsets)...)
+	encoder.appendUint(uint64(b.writtenLen), idFileSize)
```
Comment (coocood) [inline table/builder.go]: Is `idFileSize` the same as the last block end offset?

File: table/builder.go [hunk 4]
```diff
@@ -424,3 +437,69 @@ func (b *Builder) compressData(data []byte) ([]byte, error) {
...
+	y.Assert(e.buf[e.cursor]&3 == typeBytes)
+	e.cursor += 1
+	l := int(bytesToU32(e.buf[e.cursor:]))
+	e.cursor += 4
+	d := e.buf[e.cursor : e.cursor+l]
+	e.cursor += l
+	return d
+}
+
+func (e *metaDecoder) skip() {
```
Comment (coocood) [inline table/builder.go]: Or we can just encode everything in bytes.

Comment (bobotu) [reply table/builder.go]: Yes, the penalty is extra 4 bytes for fixed-length type.

File: table/builder.go [hunk 5]
```diff
@@ -307,67 +323,54 @@ func (b *Builder) Finish() error {
...
-
 	// Don't compress the global ts, because it may be updated during ingest.
 	ts := uint64(math.MaxUint64)
 	if b.isExternal {
 		// External builder doesn't append ts to the keys, the output sst should has a non-MaxUint64 global ts.
 		ts = math.MaxUint64 - 1
 	}
-	binary.BigEndian.PutUint64(encodeBuf[:], ts)
-	if err := b.w.Append(encodeBuf[:]); err != nil {
+	if err := b.w.Append(u64ToBytes(ts)); err != nil {
```
Comment (coocood) [inline table/builder.go]: `ts` doesn't have an id?

File: table/builder.go (hunk 3 above)
Comment (bobotu) [reply table/builder.go]: Yes, but I think to save it explicitly is easier to understanding

File: table/builder.go (hunk 5 above)
Comment (bobotu) [reply table/builder.go]: No, it is stored separately at the beginning of the index file. This field may be updated directly when doing ingest.

Comment (bobotu) [reply table/builder.go]: Others are stored after `ts` and are compressed

**Comment by coocood** [comment]
LGTM

**Pull request merged**